udpate: add support for SelectOnValue, find item in the tree based on…

tophe opened the pull request:
add support for SelectOnValue, find trie elements based on content and an eval function. Example usage in test

Commit:
udpate: add support for SelectOnValue, find item in the tree based on Value, using eval function

Commit:
update: renaming, cleaning tests

shivamMg commented:
Hey @tophe thanks for creating this pr.

I feel your changes are pretty close to the "Walk()" feature found in most tree implementations e.g. in [filepath.Walk()](https://pkg.go.dev/path/filepath#Walk). I propose changes that keep it similar to this feature:

```go
type WalkFunc func(key string, value interface{}) error

func (t *Trie) Walk(walker WalkFunc) error {}
```

Usage of this function would look like this:

```go
	tri := trie.New()
	tri.Put([]string{"d", "a", "l", "i"}, 2)
	tri.Put([]string{"d", "a", "l", "i", "b"}, 1)
	tri.Put([]string{"d", "a", "l", "i", "b", "e"}, 2)
	tri.Put([]string{"d", "a", "l", "i", "b", "e", "r", "t"}, 1)

	err := tri.Walk(func(key string, value interface{}) error {
		what := value.(int)
		if what == 2 {
			return errors.New("found")
		}
		return nil
	})
        if err.String() == "found" {
               println("found")
        }
```

shivamMg commented on go.mod:
```diff
@@ -1,4 +1,4 @@
-module github.com/shivamMg/trie
+module github.com/tophe/trie
```
nit: remove

shivamMg commented on search.go:
```diff
@@ -353,6 +353,46 @@ func (t *Trie) search(prefixKey []string, opts *SearchOptions) *SearchResults {
+// selectOnValue: recursively apply eval an item tree
+func (t *Trie) selectOnValue(results *SearchResults, node *Node, prefixKey []string, eval func(interface{}) bool) *SearchResults {
+
+	for key, child := range node.children {
```
Use childrenDLL to give it a deterministic walk-order. For reference see: https://github.com/shivamMg/trie/blob/master/search.go#L196
![image](https://user-images.githubusercontent.com/7882752/223314163-6317c580-e226-477d-ae42-ba2556e6a691.png)

Idea is to iterate through children using a doubly-linked list instead of map keys which is not deterministic

tophe replied:
I wonder, why you where using childrenDLL and children attributes.
I have renamed the function walk, and change the implementation tu use the childrenDLL. test are updated to.

Commit:
update: renaming, SelectOnValue -> Walk, implement with childrenDLL to avoid non deterministic sort.

tophe commented:
I have done the requested changes. walk on the way.
the match function is realy fast !

shivamMg commented on search.go:
```diff
@@ -353,6 +353,67 @@ func (t *Trie) search(prefixKey []string, opts *SearchOptions) *SearchResults {
+	return results
+}
+
+func (t *Trie) olddwalk(results *SearchResults, node *Node, prefixKey []string, walker WalkFunc) *SearchResults {
```
remove olddwalk since it's not used

shivamMg commented on search.go:
```diff
@@ -353,6 +353,67 @@ func (t *Trie) search(prefixKey []string, opts *SearchOptions) *SearchResults {
+		pfx := append(prefixKey, key)
+		if node.isTerminal {
+			if walker(node.value) {
+				t.scanBuild(results, node, &pfx)
```
no need for scanBuild, since we don't want to return *SearchResults

shivamMg commented on trie_test.go:
```diff
@@ -8,6 +8,59 @@ import (
 
+func TestTrie_Scan(t *testing.T) {
+	tri := trie.New()
+	tri.Put([]string{"d", "a", "l", "i"}, 2)
```
Test for Walk() should assert that all (key, value) pairs were returned in-order.

tophe commented:
I Shivam, I 've just  understand what who expected about the walk function, it's a map function rooted at the key node.
That's interesting, because it can update nodes based on the content, not  the key, starting at any level (defaulted to root).
That's one part off my need because I have to live udpate a 1M value trie, but I also need to scan the trie on content and return a SearchResult of the scan.

The original need for me, witch was in 
`
type SelectFunc func(interface{}) bool
func (t *Trie)  SelectOnValue(select SelectFunc) *SearchResults 
`
the idea is to return a result build on the content of the node, sometime you don't have the key but a value to find in the tree, you have to walk all the tree value to search in the content with the select function.
if the select func is true for a node you add it to the SearchResult.

In my use case I have to live update a trie based on the content of the trie not the key. So I used SelectOnValue to find all the nodes to update. Then I can update the content (delete or update one data in the slice of the value) using trie.put. 
the search is based upon the content not the key, you have to find the key matching select func.

with the walk function
`
type WalkFunc func(key string, value interface{}) error
func (t *Trie) Walk(walker WalkFunc) error {}
`
I can update a tree on based on content and so maintain the live content, but I can't Select on content, to extract matching keys.

Perhaps, we can add a Map not rooted Walk rooted and a Select function not rooted. 
if you agree with that proposition, I can work on it.

tophe commented:
hi, there is a branch on my cloned repo (walk), the filter function work, but the walk don't modify the trie. I have made different test, with value/pointer, the trie value isn't updated. tests failed. Any ideas.

shivamMg commented:
Thanks for the clarification. I think I understood your requirement (update/delete based on values). I still feel it can be done using Walk. Could you confirm the following will solve your requirements:

```go
var keysToDelete [][]string
tri.Walk(func (key []string, node *Node) error {
  if node.Value() == 2 {
    node.SetValue(3)  // new method: lets you set values
  }
  if node.Value() == 7 {
    keysToDelete = append(keysToDelete, key)
  }
  return nil
})

for _, key := range keysToDelete {
  tri.Delete(key)
}
```

tophe commented:
Hi Shivam, 

I think walk func should only update, deleting should be done with the trie.Delete method.

To do delete based upon content, we need a (filter / select / scan) method that return a SearchResult.
that's the purpose of the original PR with the SelectedOnValue func, now renamed to filter.

When you have the searchResult, you can udpate via the trie.Put or delete via the trie.Delete.

Walk is a cool addition but filer/select/scan is a mandatory func to maintain the trie based upon is content and to search in the trie based upon it's content. 

I used it for update and for search on content.

I updated my branch/walk, it just contain the Filter function, perhaps Scan can be a better name. 
I can't add the Walk func, because it need the SetValue func.

Commit:
Add support for Trie.Walk() and Node.SetValue()

Commit:
Fix err on Windows + Fix docs

shivamMg commented:
@tophe I've extended your work with what I proposed. You can try it with `feature/walk` branch.
Here's your modified test: https://github.com/shivamMg/trie/pull/4/files#diff-e22c9276175f0dbe6684cac895d7ff7303b79f998daa092aa06295ca3183443e

Do let me know if it looks good.

Commit:
update: Walk tests.

Commit:
update: include Filter

Commit:
Merge branch 'master' into feature/walk

shivamMg commented:
Closing this as most changes were cherry-picked into: https://github.com/shivamMg/trie/pull/4

@tophe Thank you for your contribution and inputs.

shivamMg commented:
https://github.com/shivamMg/trie/releases/tag/v0.0.1
